pkg/util/ticker.go

```diff
@@ -0,0 +1,65 @@
+package util
+
+import (
+	"context"
+	"math/rand"
+	"time"
+)
+
+type Jitter struct {
+	base      time.Duration
+	deviation time.Duration
+}
+
+// NewJitter returns a Jitter object that creates durations with random jitter.
+func NewJitter(b time.Duration, d time.Duration) Jitter {
+	return Jitter{base: b, deviation: d}
+}
+
+// Duration returns a random duration from the base duration and +/- jitter
+func (j Jitter) Duration() time.Duration {
+	base := j.base - j.deviation
+	jitter := time.Duration(rand.Int63n(int64(float64(2 * j.deviation.Nanoseconds()))))
+	return base + jitter
+}
+
+type TickerWithJitter struct {
+	C    chan time.Time
+	ctx  context.Context
+	stop func()
+}
+
+func (t *TickerWithJitter) Stop() {
+	t.stop()
+}
+
+func (t *TickerWithJitter) start(d, dev time.Duration) {
+	j := NewJitter(d, dev)
+	timer := time.NewTimer(j.Duration())
+	defer timer.Stop()
+	for {
+		select {
+		case <-timer.C:
+			timer.Reset(j.Duration())
+			t.C <- time.Now()
+		case <-t.ctx.Done():
+			if !timer.Stop() {
+				<-timer.C
```

Comment:
could this block if the channel is empty?

Reply:
From the Go docs:

```
// To ensure the channel is empty after a call to Stop, check the
// return value and drain the channel.
// For example, assuming the program has not received from t.C already:
//
//	if !t.Stop() {
//		<-t.C
//	}
//
// This cannot be done concurrent to other receives from the Timer's
// channel or other calls to the Timer's Stop method.
```

Since I don't receive from the timer's channel concurrently - but sequentially in the `for-select` loop - the timer can fire, then it receives from the channel, or the context is done, then the timer hasn't fired yet and the timer channel needs to be drained.

Reply:
thanks @chaudum 
comment on `Reset` also recommends doing this
https://cs.opensource.google/go/go/+/refs/tags/go1.20.6:src/time/sleep.go;l=109